microsoft/fhir-server
# _revinclude support

This is a PR that adds support for the _revinclude functionality for all spec versions (under 'shared').
It is the first phase of a two phases work, where in the next phase we will add limit to the number of included items (currently unbounded).


**update:**
added phase 2: added a limit to the number of included items per match.
this required a significant sql change. 
To verify I ran:

https://localhost:44348/Patient?_revinclude=Observation:subject

And verified the returned sql statement is as expected.
To ease review please take a look at the below expected generated query:

exec sp_executesql N'SET STATISTICS IO ON;
SET STATISTICS TIME ON;

WITH cte0 AS
(
    SELECT ResourceSurrogateId AS Sid1
    FROM dbo.Resource
    WHERE IsHistory = 0
        AND IsDeleted = 0
        AND ResourceTypeId = @p0
    
),
cte1 AS
(
    SELECT DISTINCT TOP (@p1) Sid1, 1 AS IsMatch, 0 AS IsPartial 
    FROM cte0
    ORDER BY Sid1 ASC
),
cte2 AS
(
    SELECT refSource.ResourceSurrogateId AS Sid1 
    , (CASE WHEN CountSource > 100 THEN refTarget.ResourceSurrogateId ELSE NULL END) IsPartialTargetSid1 FROM 
    dbo.Resource refTarget
    CROSS APPLY -- we want the top to work on the (rev)included resource records and not on the overall resultset, hence can''t use a simple join.
    ( 
    SELECT DISTINCT TOP (100) ResourceSurrogateId
    , count(*) over() as CountSource -- count all records before the TOP
    FROM 
    dbo.ReferenceSearchParam refSource
    WHERE refSource.ReferenceResourceTypeId = refTarget.ResourceTypeId
        AND refSource.ReferenceResourceId = refTarget.ResourceId
        AND refSource.IsHistory = 0
    ) AS 
    refSource WHERE refTarget.IsHistory = 0
        AND refTarget.ResourceTypeId = @p0
        AND refTarget.ResourceSurrogateId IN (SELECT Sid1 FROM cte1)
),
cte3 AS
(
    SELECT DISTINCT 
    cte1.Sid1 , 1 AS IsMatch , (CASE WHEN IsPartialTargetSid1 IS NULL THEN 0 ELSE 1 END) as IsPartial -- if cte2 includes the parent sid1 it means that results have been truncated
     From 
    cte1 LEFT OUTER JOIN 
    cte2 ON cte1.Sid1 = cte2.IsPartialTargetSid1 UNION ALL
     SELECT Sid1, 0 AS IsMatch, 0 AS IsPartial
     FROM 
    cte2)
SELECT r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(IsMatch AS bit) AS IsMatch, CAST(IsPartial AS bit) AS IsPartial, r.RawResource
FROM dbo.Resource r
INNER JOIN cte3
ON r.ResourceSurrogateId = cte3.Sid1
ORDER BY r.ResourceSurrogateId ASC
OPTION(RECOMPILE)',N'@p0 smallint,@p1 int',@p0=103,@p1=11
